<a href="https://colab.research.google.com/github/AmirNBA9/MachineLearning/blob/master/KNN_suggestionbook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [46]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [47]:
import pandas as pd
import numpy as np

users = pd.read_csv('/content/drive/MyDrive/ML/BX-Users.csv',on_bad_lines='skip', encoding= 'latin-1').reset_index()
books = pd.read_csv('/content/drive/MyDrive/ML/BX-Books.csv',on_bad_lines='skip', encoding= 'latin-1').reset_index()
ratings = pd.read_csv("/content/drive/MyDrive/ML/BX-Book-Ratings.csv",on_bad_lines='skip', encoding= 'latin-1').reset_index()

users.info()
books.info()
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272233 entries, 0 to 272232
Data columns (total 3 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   level_0                   272232 non-null  object
 1   level_1                   272231 non-null  object
 2   User-ID;"Location";"Age"  272230 non-null  object
dtypes: object(3)
memory usage: 6.2+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 228148 entries, 0 to 228147
Data columns (total 2 columns):
 #   Column                                                                                                       Non-Null Count   Dtype 
---  ------                                                                                                       --------------   ----- 
 0   index                                                                                                        228148 non-null  int64 
 1   ISBN;"Book-Title";"Book-Author";"Year-Of-Publi

**بررسي ديتا**
##كاربرها
- RowCount: 272230 not null data
- Structure: User-ID;"Location";"Age"

##كتاب ها
- RowCount: 228148 not null data
- Structure: ISBN;"Book-Title";"Book-Author";"Year-Of-Publication";"Publisher";"Image-URL-S";"Image-URL-M";"Image-URL-L"

##امتيازات كتاب ها
- RowCount: 1149775 not null record
- Structure: User-ID;"ISBN";"Book-Rating"

# مطالعه بيشتر ديتاها

In [49]:
# Join ratings and books dataframes
rating_users = pd.merge(ratings,users,on='ISBN')

# Shape of the data
rating_books.shape

KeyError: ignored

In [ ]:
# Take 1 % data as sample
rating_books_sample = rating_books.sample(frac=.01, random_state=1)

# Shape of the sample data
rating_books_sample.shape

#KNN

In [ ]:
# Import NearestNeighbors
from sklearn.neighbors import NearestNeighbors

# Build NearestNeighbors Object
model_nn = NearestNeighbors(metric='cosine', algorithm='brute', n_neighbors=7, n_jobs=-1)

# Fit the NearestNeighbor
model_nn.fit(rating_books_pivot)

In [ ]:
# Get top 10 nearest neighbors
indices=model_nn.kneighbors(rating_books_pivot.loc[['10 Secrets for Success and Inner Peace']], 10, return_distance=False)

# Print the recommended books
print("Recommended Books:")
print("==================")
for index, value in enumerate(rating_books_pivot.index[indices][0]):
    print((index+1),". ",value)